src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,488 @@
+            norm_cast at hle,
+            exact hle },
+          { have : 1 / s E < bdd k,
+            { linarith [le_of_max_le_left (hk k le_rfl)] {restrict_type := ℝ} },
+            rw one_div at this ⊢,
+            rwa inv_lt (lt_trans (inv_pos.2 hE₃) this) hE₃ } },
+        obtain ⟨k, hk₁, hk₂⟩ := this,
+        have hA' : A ⊆ i \ ⋃ l ≤ k, restrict_nonpos_seq s i l,
+        { rw hA,
+          apply set.diff_subset_diff_right,
+          intro x, simp only [set.mem_Union],
+          rintro ⟨n, _, hn₂⟩,
+          exact ⟨n, hn₂⟩ },
+        refine find_exists_one_div_lt_min (hn' k)
+          (buffer.lt_aux_2 hk₁) ⟨E, set.subset.trans hE₂ hA', hE₁, _⟩,
+        convert hk₂, norm_cast,
+        exact nat.sub_add_cancel hk₁ },
+      { exact lt_of_le_of_lt h₂ hi } },
+    { exact exists_subset_restrict_nonpos' hi₁ hi hn } }
+end .
+
+end exists_subset_restrict_nonpos
+
+/-- The set of measures of the set of measurable negative sets. -/
+def measure_of_negatives (s : signed_measure α) : set ℝ :=
+  s '' { B | measurable_set B ∧ s ≤[B] 0 }
+
+lemma zero_mem_measure_of_negatives : (0 : ℝ) ∈ s.measure_of_negatives :=
+⟨∅, ⟨measurable_set.empty, restrict_empty_le_zero _⟩, s.empty⟩
+
+lemma bdd_below_measure_of_negatives :
+  bdd_below s.measure_of_negatives :=
+begin
+  simp_rw [bdd_below, set.nonempty, lower_bounds, set.mem_set_of_eq],
+  by_contra, push_neg at h,
+  have h' : ∀ n : ℕ, ∃ y : ℝ, y ∈ s.measure_of_negatives ∧ y < -n := λ n, h (-n),
+  choose f hf using h',
+  have hf' : ∀ n : ℕ, ∃ B ∈ { B | measurable_set B ∧ s ≤[B] 0 }, s B < -n,
+  { intro n,
+    rcases hf n with ⟨⟨B, hB₁, hB₂⟩, hlt⟩,
+    exact ⟨B, hB₁, hB₂.symm ▸ hlt⟩ },
+  choose B hB using hf',
+  have hmeas : ∀ n, measurable_set (B n) := λ n, let ⟨h, _⟩ := (hB n).1 in h,
+  set A := ⋃ n, B n with hA,
+  have hfalse : ∀ n : ℕ, s A ≤ -n,
+  { intro n,
+    refine le_trans _ (le_of_lt (hB n).2),
+    rw [hA, ← set.diff_union_of_subset (set.subset_Union _ n),
+        of_union (disjoint.comm.1 set.disjoint_diff) _ (hmeas n)],
+    { refine add_le_of_nonpos_left _,
+      have : s ≤[A] 0 :=
+        restrict_le_restrict_Union _ _ hmeas (λ m, let ⟨_, h⟩ := (hB m).1 in h),
+      refine nonpos_of_restrict_le_zero _ (restrict_le_zero_subset _ _ this (set.diff_subset _ _)),
+      exact measurable_set.Union hmeas },
+    { apply_instance },
+    { exact (measurable_set.Union hmeas).diff (hmeas n) } },
+  suffices : ¬ ∀ n : ℕ, s A ≤ -n,
+  { exact this hfalse },
+  push_neg,
+  rcases exists_nat_gt (-(s A)) with ⟨n, hn⟩,
+  exact ⟨n, neg_lt.1 hn⟩,
+end
+
+/-- **The Hahn decomposition thoerem**: Given a signed measure `s`, there exist
+disjoint measurable sets `i`, `j` such that `i` is positive, `j` is negative
+and `i ∪ j = set.univ`.  -/
+theorem exists_disjoint_positive_negative_union_eq (s : signed_measure α) :
+  ∃ i j : set α, measurable_set i ∧ 0 ≤[i] s ∧ measurable_set j ∧ s ≤[j] 0 ∧
+  disjoint i j ∧ i ∪ j = set.univ :=
```

Comment:
Isn't the way to spell "disjoint and union is `univ`" `is_compl`?
```suggestion
  ∃ i j : set α, measurable_set i ∧ 0 ≤[i] s ∧ measurable_set j ∧ s ≤[j] 0 ∧
  is_compl i j :=
```

Reply:
Which suggests to me that maybe this statement should be
```suggestion
  ∃ i : set α, measurable_set i ∧ 0 ≤[i] s ∧ s ≤[iᶜ] 0 :=
```

Reply:
Oh, I see you have that statement below! I think it would be easier to prove `exists_compl_positive_negative` in terms of `exists_disjoint_positive_negative_union_eq` rather than vice versa.

Reply:
I think in general the second version is almost always more useful (I created the second version since I kept finding myself using it), so perhaps it make sense to not include the first version at all

Reply:
That seems sensible to me.